src/iceberg/update/update_partition_spec.cc

```diff
@@ -0,0 +1,488 @@
+UpdatePartitionSpec::UpdatePartitionSpec(std::shared_ptr<Transaction> transaction)
+    : PendingUpdate(std::move(transaction)) {
+  const TableMetadata* base_metadata = transaction_->base();
+  if (base_metadata == nullptr) [[unlikely]] {
+    AddError(ErrorKind::kInvalidArgument,
+             "Base table metadata is required to construct UpdatePartitionSpec");
+    return;
+  }
+  format_version_ = base_metadata->format_version;
+
+  // Get the current/default partition spec
+  auto spec_result = base_metadata->PartitionSpec();
+  if (!spec_result.has_value()) {
+    AddError(spec_result.error());
+    return;
+  }
+  spec_ = std::move(spec_result.value());
+
+  // Get the current schema
+  auto schema_result = base_metadata->Schema();
+  if (!schema_result.has_value()) {
+    AddError(schema_result.error());
+    return;
+  }
+  schema_ = std::move(schema_result.value());
+
+  last_assigned_partition_id_ = spec_->last_assigned_field_id();
+  name_to_field_ = IndexSpecByName(*spec_);
+  transform_to_field_ = IndexSpecByTransform(*spec_);
+
+  // Check for unknown transforms
+  for (const auto& field : spec_->fields()) {
+    if (field.transform()->transform_type() == TransformType::kUnknown) {
+      AddError(ErrorKind::kInvalidArgument,
+               "Cannot update partition spec with unknown transform: {}",
+               field.ToString());
+      return;
+    }
+  }
+
+  // Build index of historical partition fields for efficient recycling (V2+)
+  if (format_version_ >= 2) {
+    BuildHistoricalFieldsIndex();
+  }
+}
+
+UpdatePartitionSpec::~UpdatePartitionSpec() = default;
+
+UpdatePartitionSpec& UpdatePartitionSpec::CaseSensitive(bool is_case_sensitive) {
+  case_sensitive_ = is_case_sensitive;
+  return *this;
+}
+
+UpdatePartitionSpec& UpdatePartitionSpec::AddNonDefaultSpec() {
+  set_as_default_ = false;
+  return *this;
+}
+
+UpdatePartitionSpec& UpdatePartitionSpec::AddField(const std::string& source_name) {
+  // Find the source field in the schema
+  ICEBERG_BUILDER_ASSIGN_OR_RETURN(
+      auto field_opt, schema_->FindFieldByName(source_name, case_sensitive_));
+
+  ICEBERG_BUILDER_CHECK(field_opt.has_value(), "Cannot find source field: {}",
+                        source_name);
+  int32_t source_id = field_opt->get().field_id();
+  return AddFieldInternal(nullptr, source_id, Transform::Identity());
+}
+
+UpdatePartitionSpec& UpdatePartitionSpec::AddField(std::shared_ptr<NamedReference> term,
+                                                   std::string part_name) {
+  // Bind the term to get the source field
+  ICEBERG_BUILDER_ASSIGN_OR_RETURN(auto bound_ref, term->Bind(*schema_, case_sensitive_));
+  int32_t source_id = bound_ref->field().field_id();
+
+  // Reference terms use identity transform
+  return AddFieldInternal(part_name.empty() ? nullptr : &part_name, source_id,
+                          Transform::Identity());
+}
+
+UpdatePartitionSpec& UpdatePartitionSpec::AddField(std::shared_ptr<UnboundTransform> term,
```

Comment:
ditto

Reply:
BTW, should we consolidate this with the above function using the signature below?

```cpp
AddField(const std::shared_ptr<Term>& term, const std::string& part_name)
```

We need to check `term->is_unbound()` and `term->kind()` to cast to the right type.